### PR TITLE
Writing to Hive bucket table

### PR DESCRIPTION
**Pull request by haozhun**

### CODE REVIEW

**Comment by dain** [comment]
@electrum the first commit has this comment:
```
* Fail instead of fallback when invalid bucket data is detected
  * It is not possible to retain the old behavior after implementing table
    distribution in Hive connector
```
Do you have any issues with this?

File: presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java [hunk 1]
```diff
@@ -296,66 +293,53 @@ private void loadPartition(HivePartitionMetadata partition)
...
-            return Optional.empty();
-        }
-
-        Map<String, FileStatus> map = new HashMap<>();
-        List<String> paths = new ArrayList<>();
-        for (FileStatus status : statuses) {
-            if (!isFile(status)) {
-                return Optional.empty();
+        while (hiveFileIterator.hasNext()) {
+            if (list.size() >= bucketCount) {
```
Comment (dain) [inline presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java]: Place this check after the directory check, so you get the directory message if the last file is a directory.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveBucketing.java [hunk 2]
```diff
@@ -112,7 +119,14 @@ private static int hash(TypeInfo type, Block block, int position)
                     case BOOLEAN:
                         return BOOLEAN.getBoolean(block, position) ? 1 : 0;
                     case BYTE:
+                        // Check that the value is within range of the specified type.
+                        // * For reading, this should never fail.
+                        // * For writing, we don't support writing to BYTE, SHORT, INT, FLOAT right now.
```
Comment (dain) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveBucketing.java]: I believe that @raghavsethi is about to check in the INT code

File: presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java [hunk 3]
```diff
@@ -310,10 +321,38 @@ private void loadPartition(HivePartitionMetadata partition)
...
             return;
         }
 
+        // If table is bucketed: list the directory, sort, tag with bucket id
+        if (bucketHandle.isPresent()) {
+            // HiveFileIterator skips hidden files automatically.
+            int bucketCount = bucketHandle.get().getBucketCount();
+            List<LocatedFileStatus> list = listAndSortBucketFiles(iterator, bucketCount);
+
+            for (int i = 0; i < bucketCount; i++) {
```
Comment (dain) [inline presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java]: rename `i` to `bucketNumber` or `bucket`

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveBucketFunction.java [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+        this.typeInfos = requireNonNull(hiveTypes, "hiveTypes is null").stream()
+                .map(HiveType::getTypeInfo)
+                .collect(Collectors.toList());
+    }
+
+    @Override
+    public int getBucket(Page page, int position)
+    {
+        Block block = page.getBlock(0);
+        if (block.isNull(position)) {
```
Comment (dain) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveBucketFunction.java]: why check only the first block for null?

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveBucketProperty.java [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+
+public class HiveBucketProperty
+{
+    private final List<String> clusteredBy;
+    private final int bucketCount;
+
+    @JsonCreator
+    public HiveBucketProperty(@JsonProperty("clusteredBy") List<String> clusteredBy, @JsonProperty("bucketCount") int bucketCount)
+    {
+        this.clusteredBy = clusteredBy;
```
Comment (dain) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveBucketProperty.java]: check null

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveBucketHandle.java [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+
+public class HiveBucketHandle
+{
+    private final List<HiveColumnHandle> columns;
+    private final int bucketCount;
+
+    @JsonCreator
+    public HiveBucketHandle(@JsonProperty("columns") List<HiveColumnHandle> columns, @JsonProperty("bucketCount") int bucketCount)
+    {
+        this.columns = columns;
```
Comment (dain) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveBucketHandle.java]: check null

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveBucketProperty.java [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+        this.bucketCount = bucketCount;
+    }
+
+    public static Optional<HiveBucketProperty> fromStorageDescriptor(StorageDescriptor storageDescriptor, String tablePartitionName)
+    {
+        boolean bucketColsSet = storageDescriptor.isSetBucketCols() && !storageDescriptor.getBucketCols().isEmpty();
+        boolean numBucketsSet = storageDescriptor.isSetNumBuckets() && storageDescriptor.getNumBuckets() > 0;
+        if (!bucketColsSet || !numBucketsSet) {
+            if (bucketColsSet || numBucketsSet) {
```
Comment (dain) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveBucketProperty.java]: I would put this first as `if (bucketColsSet != numBucketsSet)` and the the remain `if` structure is much simpler

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java [hunk 8]
```diff
@@ -835,7 +852,11 @@ private Partition createPartition(Table table, PartitionUpdate partitionUpdate)
             partition.getSd().setLocation(partitionUpdate.getTargetPath());
         }
         else {
-            partition.setSd(makeStorageDescriptor(table.getTableName(), defaultStorageFormat, new Path(partitionUpdate.getTargetPath()), table.getSd().getCols()));
+            Optional<HiveBucketProperty> bucketProperty = Optional.empty();
+            if (table.getSd().isSetBucketCols()) {
+                bucketProperty = Optional.of(new HiveBucketProperty(table.getSd().getBucketCols(), table.getSd().getNumBuckets()));
```
Comment (dain) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java]: Use `HiveBucketProperty.fromStorageDescriptor`

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveNodePartitioningProvider.java [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+        HivePartitioningHandle handle = checkType(partitioningHandle, HivePartitioningHandle.class, "partitioningHandle");
+        List<HiveType> hiveTypes = handle.getHiveTypes();
+        return new HiveBucketFunction(bucketCount, hiveTypes);
+    }
+
+    @Override
+    public Map<Integer, Node> getBucketToNode(ConnectorTransactionHandle transactionHandle, ConnectorSession session, ConnectorPartitioningHandle partitioningHandle)
+    {
+        List<Node> nodes = ImmutableList.copyOf(nodeManager.getActiveDatasourceNodes(connectorId));
+        checkState(!nodes.isEmpty(), "No %s nodes available", connectorId);
```
Comment (dain) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveNodePartitioningProvider.java]: I'd shuffle the nodes also, just to be safe

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveOutputTableHandle.java [hunk 10]
```diff
@@ -51,7 +53,8 @@ public HiveOutputTableHandle(
                 tableName,
                 inputColumns,
                 filePrefix,
-                requireNonNull(locationHandle, "locationHandle is null"),
+                locationHandle,
```
Comment (dain) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveOutputTableHandle.java]: Why remove the null check?

File: presto-hive/src/main/java/com/facebook/presto/hive/HivePageSink.java [hunk 11]
```diff
@@ -198,11 +221,35 @@ public HivePageSink(
...
+                dataColumnNameToTypeMap.put(column.getName(), column.getHiveType());
             }
         }
         this.partitionColumnsInputIndex = Ints.toArray(partitionColumns.build());
         this.dataColumnInputIndex = Ints.toArray(dataColumnsInputIndex.build());
 
+        requireNonNull(bucketProperty, "bucketProperty is null");
+        if (bucketProperty.isPresent()) {
+            int bucketCount = bucketProperty.get().getBucketCount();
+            checkArgument(bucketCount < MAX_BUCKET_COUNT, "bucketCount must be smaller than 100000");
```
Comment (dain) [inline presto-hive/src/main/java/com/facebook/presto/hive/HivePageSink.java]: you should check this during generation of the bucket property also

File: presto-hive/src/main/java/com/facebook/presto/hive/HivePageSink.java [hunk 12]
```diff
@@ -263,28 +330,62 @@ public void abort()
...
+                HiveRecordWriter writer = writers.get(bucket);
+                if (writer == null) {
+                    for (int field = 0; field < partitionBlocks.length; field++) {
+                        Object value = getField(partitionColumnTypes.get(field), partitionBlocks[field], position);
+                        partitionRow.set(field, value);
+                    }
+                    writer = createWriter(partitionRow, filePrefix + "_bucket-" + Strings.padStart(Integer.toString(bucket), BUCKET_NUMBER_PADDING, '0'));
+                    writers.put(bucket, writer);
+                    bucketWriterCount++;
+                    if (bucketWriterCount >= maxOpenPartitions) {
```
Comment (dain) [inline presto-hive/src/main/java/com/facebook/presto/hive/HivePageSink.java]: Maybe put this check before opening the writer

File: presto-hive/src/main/java/com/facebook/presto/hive/HivePartitionManager.java [hunk 13]
```diff
@@ -97,13 +98,15 @@ public HivePartitionResult getPartitions(ConnectorSession session, HiveMetastore
...
+        Optional<HiveBucketHandle> hiveBucketHandle = getHiveBucketHandle(connectorId, table);
+
         List<HiveColumnHandle> partitionColumns = getPartitionKeyColumnHandles(connectorId, table);
         Optional<HiveBucketing.HiveBucket> bucket = getHiveBucket(table, TupleDomain.extractFixedValues(effectivePredicate).get());
 
         TupleDomain<HiveColumnHandle> compactEffectivePredicate = toCompactTupleDomain(effectivePredicate, domainCompactionThreshold);
 
         if (effectivePredicate.isNone()) {
-            return new HivePartitionResult(partitionColumns, ImmutableList.of(), TupleDomain.none(), TupleDomain.none());
+            return new HivePartitionResult(partitionColumns, ImmutableList.of(), effectivePredicate, TupleDomain.none(), hiveBucketHandle);
```
Comment (dain) [inline presto-hive/src/main/java/com/facebook/presto/hive/HivePartitionManager.java]: Why the change to `effectivePredicated` here?

File: presto-hive/src/main/java/com/facebook/presto/hive/HivePartitioningHandle.java [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+public class HivePartitioningHandle
+        implements ConnectorPartitioningHandle
+{
+    private final String clientId;
+    private final int bucketCount;
+    private final List<HiveType> hiveTypes;
+
+    @JsonCreator
+    public HivePartitioningHandle(
+            @JsonProperty("clientId") String clientId,
```
Comment (dain) [inline presto-hive/src/main/java/com/facebook/presto/hive/HivePartitioningHandle.java]: @electrum do we still need `clientId` in Hive handles?

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java [hunk 15]
```diff
@@ -294,6 +296,71 @@ public void testCreateTableNonExistentPartitionColumns()
...
+        assertEquals(tableMetadata.getMetadata().getProperties().get(BUCKET_COUNT_PROPERTY), 11);
+
+        List<HivePartition> partitions = getPartitions("test_create_partitioned_bucketed_table_as");
+        assertEquals(partitions.size(), 3);
+
+        // Hive will reorder the partition keys to the end
+        assertQuery("SELECT * from test_create_partitioned_bucketed_table_as", "SELECT custkey, comment, orderstatus FROM orders");
+
+        try {
+            assertUpdate("INSERT INTO test_create_partitioned_bucketed_table_as VALUES (1, 'comment', 'O')", 1);
```
Comment (dain) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java]: Add a test that does a successful insert into a new partition and read some data from it.

**Comment by dain** [comment]
Some small comments and questions, but overall looks good.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveOutputTableHandle.java (hunk 10 above)
Comment (haozhun) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveOutputTableHandle.java]: This null check shouldn't have been added in the first place because the other arguments here don't have null test. And super will do the null test.

File: presto-hive/src/main/java/com/facebook/presto/hive/HivePageSink.java (hunk 11 above)
Comment (haozhun) [inline presto-hive/src/main/java/com/facebook/presto/hive/HivePageSink.java]: There isn't a particular reason why there must not be more buckets than 100,000 for reading.

**Comment by haozhun** [comment]
Addressed comments from @dain.

**Comment by electrum** [comment]
@dain That seems fine to me. It's arguably better to fail when we detect that a table is broken.

File: presto-hive/src/main/java/com/facebook/presto/hive/HivePartitioningHandle.java (hunk 14 above)
Comment (electrum) [inline presto-hive/src/main/java/com/facebook/presto/hive/HivePartitioningHandle.java]: Nope, you can remove that everywhere!

**Comment by dain** [comment]
There are some build failures due to unused imports.  Assuming the necessary changes are small and the tests pass, this is good to merge.

You can make the change to remove the `clientId` from `HivePartitioningHandle` in another PR.  Also, you need to sync with @raghavsethi on the new types code to make sure it works with bucketed writes correctly.
